src/pages/index.js

```diff
@@ -84,26 +104,153 @@ editCardPopup.setEventListeners();
 const userInfo = new UserInfo({
   titleSelector: ".profile__title",
   descriptionSelector: ".profile__description",
+  avatar: ".profile__image",
 });
 
+/*Delete card confirmation*/
+
+const deleteConfirmPopup = new PopupWithConfirmation({
+  popupSelector: "#modal-delete-card",
+  handleFormSubmit: () => {},
+});
+deleteConfirmPopup.setEventListeners();
+
+/*Edit Avatar*/
+
+const avatarProfilePopup = new PopupWithForm(
+  "#profile-avatar-modal",
+  avatarSubmit
+);
+avatarProfilePopup.setEventListeners();
+
+const EditAvatarBtn = document.querySelector(".profile__avatar-button");
+EditAvatarBtn.addEventListener("click", () => {
+  avatarProfilePopup.open();
+});
 /*  Event Handlers */
 
-function handleProfileEditSubmit(data) {
+function avatarSubmit(link) {
+  avatarProfilePopup.modalLoading(true);
+  api
+    .updateAvatar(link)
+    .then((res) => {
+      console.log(res);
+      userInfo.updateProfileImage(res.avatar);
+    })
+    .catch((err) => {
+      console.error(err);
+    })
+    .finally(() => {
+      avatarProfilePopup.modalLoading(false);
+    });
+  avatarProfilePopup.close();
+  avatarFormElement.reset();
+}
+
+function handleDeleteCard(card) {
+  //console.log("Delete button clicked");
+  deleteConfirmPopup.open();
+
+  deleteConfirmPopup.confirmDelete(() => {
+    // console.log("Deletion confirmed");
+    api
+
+      .deleteCard(card.getId())
+      .then(() => {
+        // console.log("Card deletion API succeeded");
+        card.removeCard();
+        deleteConfirmPopup.close();
+      })
+      .catch((err) => {
+        console.error("Card deletion API failed:", err);
+      });
+  });
+}
+
+function handleLikeIcon(card) {
+  if (card.getIsLiked()) {
+    api
+      .removeLike(card.getId())
+      .then(() => {
+        card.handleLikeCard(false);
+      })
+      .catch((err) => {
+        console.log(`Unable to process request, ${err}`);
+        // Optionally revert the UI state if there's an error
+      });
+  } else {
+    api
+      .addLike(card.getId())
+      .then(() => {
+        card.handleLikeCard(true);
+      })
+      .catch((err) => {
+        console.log(`Unable to process request, ${err}`);
+        // Optionally revert the UI state if there's an error
+      });
+  }
+}
+
+/*function handleProfileEditSubmit(data) {
   userInfo.setUserInfo(data);
   editCardPopup.close();
+}*/
+
+function handleProfileEditSubmit(userData) {
+  //console.log("User Data Submitted:", userData);
+  editCardPopup.modalLoading(true);
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```